### ISSUE TITLE
Enhance reportUnnecessaryComparison to behave like mypy's strict_equality

### ISSUE THREAD

**Issue by brianmcburney** [microsoft/pyright#3034]
It would be helpful if reportUnnecessaryComparison (or another option) reported the following problem that is reported by mypy using its strict_equality option.

The specific use case I have in mind is to handle type errors when using [sqlalchemy dataclass models](https://docs.sqlalchemy.org/en/14/orm/mapping_styles.html#imperative-mapping-with-dataclasses-and-attrs), for example when building a query:
`select(Foo.bar).where(Foo.bar == 42) # should be an error`

Here is an example:

```
from dataclasses import dataclass


@dataclass
class Foo:
    bar: str


if Foo.bar == 42:
    print("nope")
```

```
mypy --strict-equality test.py
test.py:9: error: Non-overlapping equality check (left operand type: "str", right operand type: "Literal[42]")
```

**Comment by erictraut** [comment]
Thanks for the suggestion. I agree this could be valuable.

I've been hesitant to implement a feature like this because Python allows any class to handle `__eq__` any way they'd like, and most classes that support `__eq__` annotate the method to allow any `object` as an input parameter. So a type checker can't know from the type signature of `__eq__` whether a comparison of two types should be considered "legitimate" or a potential bug.

It looks like mypy settled on [some heuristics](https://github.com/python/mypy/issues/6607) that tend to work well for most code. However, I don't know if the heuristics are good enough to enable this feature by default in strict mode. We generally don't like adding new diagnostic checks if we feel like they cannot be enabled in strict mode because such features are difficult to discover and get almost no usage.

I'll investigate this further.

**Comment by erictraut** [comment]
I've enhanced the "reportUnnecessaryComparison" diagnostic check so it catches more cases. It's not identical in implementation to the one in mypy, but it's pretty close now. This will be included in the next release.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.228, which I just published. It will also be included in the next release of pylance.

**Comment by brianmcburney** [comment]
I just tested this with the new release and noticed it isn't working for Optional types. Is that possible to support?

```
from dataclasses import dataclass
from typing import Optional

@dataclass
class Foo:
    bar: Optional[str]


if Foo.bar == 42:
    print("nope")
```

```
$ mypy --strict-equality test.py
test.py:10: error: Non-overlapping equality check (left operand type: "Optional[str]", right operand type: "Literal[42]")

$ pyright test.py
...
0 errors, 0 warnings, 0 informations 
```

**Comment by erictraut** [comment]
Oh, good catch. This was a bug in my implementation — and a hole in my unit tests. It wasn't handling `Optional` correctly. It will be fixed in the next release.

**Comment by colehaus** [comment]
Let me know if you'd prefer this to be a separate issue instead of resurrecting this one but:

I sort of would have expected `reportUnnecessaryComparison` to flag an invalid comparison with the dataclass itself but it does not. Slightly tweaking the example above:

```py
@dataclass
class Foo:
    bar: str

def foo(x: Foo):
    if x == 42: // No complaints from pyright
        print("nope")
```

Mypy with strict equality does flag this issue. I see what you're saying above about `_eq__` being pretty unconstrained, but it seems like nice functionality to have for dataclasses.

**Comment by erictraut** [comment]
@colehaus, yeah, that's reasonable. If a dataclass doesn't define its own `__eq__` method, it should be safe to assume that any other class is non-comparable. If the dataclass does define a custom `__eq__` method, then this would be an unsafe assumption. This extended support will be in the next release.

**Comment by colehaus** [comment]
Thanks!
